### ISSUE TITLE
NK Storage: Combine menu entries Generate Keys and Initialize Storage WIth Random Data

### ISSUE THREAD

**Issue by jans23** [Nitrokey/nitrokey-app#45]
When using a brand new Nitrokey Storage the App shows two menu entries:
- generate AES key
- initialize storage with random data

Both entries should be combined to "initialize device". When this function is selected the AES keys should be generated first and the mass storage be initialized afterwards.

Also the text of warning dialogs, which are displayed as long as the device is not initialized, should be updated and refer to "initialize device".

**Comment by szszszsz** [comment]
@jans23 Should these two entries always merged or only on first initialization?

**Comment by jans23** [comment]
Always.
-- 
Sent from a mobile device.

**Comment by szszszsz** [comment]
Tested on Ubuntu 16.04 with scenario:
1. Menu->Configure->Destroy encrypted data
2. Confirm generating new AES keys
3. Cancel SD card wiping
4. Check menu

'Initialize device' is visible in both cases: when device is not initialized or when sd card is not initialiazed. Warning text updated.

**Comment by szszszsz** [comment]
@jans23 Please let me know if that works for you - branch issue_45

**Comment by szszszsz** [comment]
Merged to master. Please reopen if any modifications would be needed.

**Comment by jans23** [comment]
The command "destroy encrypted data" should only trigger generating new AES keys but not initializing the storage with random data.

**Comment by szszszsz** [comment]
@jans23 Please retest and close if correct.

**Comment by jans23** [comment]
@szszszsz 
I describe the expected behaviour:

1) If brand new device is inserted, "initialize device" is displayed. User selects "initialize device", AES keys are generated and the entire storage is initialized. Finally a success message is displayed. From now on, "initialize device" should not be displayed anymore.

2) When user selects "destroy encrypted data" new AES keys are generated. **Afterwards the device appears as fully initialized. "initialize device" should not be displayed anymore.**

Reality check:

Most of it is working as expected except the sentence marked in bold. After destroying encrypted data the command "initialized device" is displayed again. Instead it shouldn't and the device should work as fully initialized. Please let me know if fixing this issue should involve a modification of the firmware.

**Comment by szszszsz** [comment]
Hi @jans23 
Thanks for clearing that out. I have another question though. From what I have observed until now creating new volumes was not possible after generating AES keys without clearing SD card. Should not clearing the card be optional?

**Comment by jans23** [comment]
Hi @szszszsz 
That behaviour doesn't make sense to me. Could you check and provide more details, please?

**Comment by szszszsz** [comment]
Currently generating new AES keys sets stick's flag `SDFillWithRandomChars_u8` to `0` on firmware side which forces GUI to block volume actions until storage is overwritten with random data. Let me rephrase your description:
- each NK Storage should be initialized only once (to generate its own AES keys)
- each SD card should be initialized only once with first use
- after AES keys regenerating SD card random overwrite should not be done (it is not needed since it is already overwritten with quasi-random data from previous encrypted use)
I will submit proper issue report to NK Storage firmware.

**Comment by szszszsz** [comment]
New patch is in branch `issue_45_3-initialize_device`. Needs test with firmware fixes, if there would be any. Need also to apply changes to device state cache on nitro-app side.

**Comment by jans23** [comment]
Describing the expected behaviour in different terms:

0. A blank device is connected
1. initialize device is displayed in the menu which:
  1. generates new AES keys
  2. formats the storage with random data
2. Once the device has been initialized successfully, the menu entry is not displayed anymore.
3. If user "destroys encrypted data" the AES keys are overwritten.

As long as the device is not flashed or reset to factory defaults the storage is not formated with random data again.

**Comment by szszszsz** [comment]
@jans23 Please test and close if all is correct.
